README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# chemdeg

<!-- badges: start -->

[![R-CMD-check](https://github.com/migliomatte/chemdeg/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/migliomatte/chemdeg/actions/workflows/R-CMD-check.yaml)
<!-- badges: end -->

chemdeg is a package developed to assist food chemists in the objective
selection of degradation kinetic models of foods and parameters
estimation. The software implements a two-step method to this purpose.
First, experimental data are analyzed in the so-called phase space which
allows for the estimation of the order of the reaction; then the data
are fitted with the appropriate standard kinetic model to estimate the
reaction rate. The whole procedure is driven by rigorous statistical
analyses of the results. The package also provides a routine to fit a
non-standard kinetic degradation model named first-order multi-target
model (FOMT).

## Installation

You can install the development version of chemdeg like so:

``` r
# from CRAN
install.packages("chemdeg")

# from GitHub
library(devtools)
devtools::install_github("migliomatte/chemdeg",
                          dependencies = TRUE, build_vignettes = TRUE)
```

## Usage

``` r
library(chemdeg)
```

#### Determination of the reaction order

The determination of the reaction order of a generic degradation kinetic
model is often prone to subjectivity. By analyzing the phase space of
the kinetic system it is possible to statistically determine the
reaction order. This information can then be used to determine the
reaction rate constant. In the package `chemdeg` the whole analysis can
be performed with the function `det_order()`:

``` r
res <- det_order(ord1)
#> Reaction order estimated: 1
```

where `ord1` is a built in function for 1<sup>st</sup>-order kinetic
model data.

The function `results` returns statistics and parameters estimates:

``` r
results(res)
#> 
#> Linear regression in the phase space: 
#> log(dx/dt)= 0.97 log(x) + ( -0.46 )
#> 
#> Estimate of n:
#> 
#>     Estimate   Std. Error      t value     Pr(>|t|) 
#> 9.701806e-01 4.235714e-03 2.290477e+02 1.835114e-07 
#> 
#> Confidence interval of n: 
#>     2.5 %    97.5 % 
#> 0.9567007 0.9836605 
#> 
#> Statistical analysis indicates that an order 1 degradation kineitc model is likely to describe the data.
#> The null hypothesis H0:
#> "The process is described by an order 1kinetic model"
#>  cannot be rejected.
#> 
#> Non-linear least squares regression was performed with an order  1  kinetic model:
#>  
#>  Estimate of k: 
#>    Estimate  Std. Error  t value     Pr(>|t|)
#> k 0.6878765 0.002626997 261.8489 1.541636e-11
#> In attesa che venga eseguita la profilazione...
#> Confidence interval of k: 
#>      2.5%     97.5% 
#> 0.6812276 0.6947116 
#> 
#> Goodness-of-fit:
#>                  Value
#> AIC:        -11.933893
#> AICc:       -10.933893
#> BIC:        -12.350374
#> RMSE:         2.002812
#> Chi-sq_red:   5.379602
#> -----------------------------------------------------
```

Plots of the data in both the phase space and the normal concentration
vs time space along with their best fits with model equations can be
accessed with `plot_ord()`:

``` r
plot_ord(res)
```

<img src="man/figures/README-unnamed-chunk-4-1.png" width="100%" /><img src="man/figures/README-unnamed-chunk-4-2.png" width="100%" />

#### First-order multi-target model fitting

The following is an example of first-order multi-target (FOMT) model.
The `FOMT()` function can be used to perform the fit:

``` r
fit_fomt <- FOMT(fomtdata)
summary(fit_fomt)
#> 
#> Formula: y ~ 1 - (1 - exp(-k * t))^n
#> 
#> Parameters:
#>   Estimate Std. Error t value Pr(>|t|)    
#> k 0.056836   0.008206   6.926 0.000449 ***
#> n 9.478174   3.926148   2.414 0.052280 .  
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
#> 
#> Residual standard error: 0.04243 on 6 degrees of freedom
#> 
#> Number of iterations to convergence: 10 
#> Achieved convergence tolerance: 3.98e-06
```

where `fomtdata` is an example data-frame provided with the package.

If the `FOMT()` function does not converge, it is possible to use the
`nls` function from `stats` package as follows:

``` r
fit_fomt1 <- nls(y ~ FOMTm(t, k, n),
  data = list(y = fomtdata$tCQA_AA, t = fomtdata$time_h),
  start = list(n = 10, k = 0.05)
)
summary(fit_fomt1)
#> 
#> Formula: y ~ FOMTm(t, k, n)
#> 
#> Parameters:
#>   Estimate Std. Error t value Pr(>|t|)    
#> n 9.478163   3.926141   2.414 0.052280 .  
#> k 0.056836   0.008206   6.926 0.000449 ***
#> ---
#> Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
#> 
#> Residual standard error: 0.04243 on 6 degrees of freedom
#> 
#> Number of iterations to convergence: 9 
#> Achieved convergence tolerance: 5.453e-06
```

#### Goodness-of-fit statistics

`chemdeg` implements the chi-squared reduced statistics
($\chi^2_{red}=\chi^2/df$ where $df$ are the degrees of freedom) and the
Akaike Information Criterion with correction for small sample size
(AICc) as goodness of fit measures. These can be accessed with the
functions `chiquad_red()`and `AICC()`, respectively.

To access the full table comprising the Bayesian Information Criterion
(BIC), Akaike Information Criterion (AIC), AICc, Root Means Squared
Error (RMSE) and $\chi^2_{red}$ (from both package `stats` and
`chemdeg`) call the function `goodness_of_fit()`:

``` r
goodness_of_fit(fit_fomt)
#>                    Value
#> AIC:        -24.15681095
#> AICc:       -21.75681095
#> BIC:        -23.91848633
#> RMSE:         0.04243006
#> Chi-sq_red:           NA
```

For more details see

``` r
vignette("chemdeg_basics","chemdeg")
```


NEWS.md
# chemdeg 0.1.2

* Debugging for `par` modifications using `on.exit()` function

# chemdeg 0.1.1

* Removing links to vignette in README file

# chemdeg 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(chemdeg)

test_check("chemdeg")


tests/testthat/test-FOMT.R
# FOMT
test_that("FOMT works with dtataframe with 3 columns(error)", {
  t <- c(0, 4, 8, 12, 16, 20)
  conc <- c(1, 0.98, 0.99, 0.67, 0.12, 0.03)
  err <- c(0.02, 0.05, 0.04, 0.04, 0.03, 0.02)
  dframe <- data.frame(t, conc, err)
  expect_no_error(FOMT <- FOMT(dframe))
  regr <- nls(conc ~ 1 - (1 - exp(-k * t))^n,
    start = par_est_FOMT(t, conc),
    weights = 1 / err^2
  )
  expect_equal(FOMT$m$getAllPars(), regr$m$getAllPars(), ignore_attr = TRUE)
})

test_that("FOMT works with dtataframe with 2 columns (no error)", {
  t <- c(0, 4, 8, 12, 16, 20)
  conc <- c(1, 0.98, 0.99, 0.67, 0.12, 0.03)
  dframe <- data.frame(t, conc)
  expect_no_error(FOMT <- FOMT(dframe))
  regr <- nls(conc ~ 1 - (1 - exp(-k * t))^n, start = par_est_FOMT(t, conc))
  expect_equal(FOMT$m$getAllPars(), regr$m$getAllPars(), ignore_attr = TRUE)
})

test_that("FOMT works with dtataframe with 2 columns (no error)", {
  t <- c(0, 4, 8, 12, 16, 20)
  conc <- c(1, 0.98, 0.99, 0.67, 0.12, 0.03)
  err <- c(0.02, 0.05, 0.04, 0.04, 0.03, 0.02)
  dframe <- data.frame(t, conc, err, err)
  expect_error(FOMT(t))
  expect_error(FOMT(dframe))
})

# FOMTm
test_that("FOMTm resturns correct values", {
  k <- 0.5
  t <- c(0, 1, 2)
  n <- 3
  expect_equal(FOMTm(k, t, n), c(1, 0.939, 0.747), tolerance = 0.001)
})

test_that("FOMTm can be used as formula in nls function", {
  t <- c(0, 4, 8, 12, 16, 20)
  conc <- c(1, 0.98, 0.99, 0.67, 0.12, 0.03)
  expect_no_error(fit <- nls(conc ~ FOMTm(k, t, n),
    start = list(k = 0.5, n = 599)
  ))
})


# par_est_FOMT
test_that("par_est_FOMT works with a data frame with 2 columns", {
  t <- c(0, 4, 8, 12, 16, 20)
  conc <- c(1, 0.98, 0.99, 0.67, 0.12, 0.03)
  dframe <- data.frame(t, conc)
  expect_no_error(par_est_FOMT(dframe))
  expect_equal(par_est_FOMT(dframe), c(k = 0.34657, n = 30.72000),
    tolerance = 1e-4
  )
})

test_that("par_est_FOMT works with a two arrays", {
  t <- c(0, 4, 8, 12, 16, 20)
  conc <- c(1, 0.98, 0.99, 0.67, 0.12, 0.03)
  expect_no_error(par_est_FOMT(t, conc))
  expect_equal(par_est_FOMT(t, conc), c(k = 0.34657, n = 30.72000),
    tolerance = 1e-4
  )
})

test_that("par_est_FOMT returns the same results with dataframe and arrays", {
  t <- c(0, 4, 8, 12, 16, 20)
  conc <- c(1, 0.98, 0.99, 0.67, 0.12, 0.03)
  dframe <- data.frame(t, conc)
  expect_no_error(par_est_FOMT(dframe))
  expect_no_error(par_est_FOMT(t, conc))
  expect_equal(par_est_FOMT(dframe), par_est_FOMT(t, conc))
})


tests/testthat/test-plot.R
test_that("plot_ord works with weighted data", {
  t <- c(0, 4, 8, 12, 16, 20)
  set.seed(1)
  conc <- (0.5 * (0.05 * t + (10^(-0.5)) / 0.5))^-2 + rnorm(length(t), 0, 0.05)
  err <- c(0.02, 0.05, 0.04, 0.04, 0.03, 0.02)
  dframe <- data.frame(t, conc, err)
  res <- det_order(dframe)
  expect_no_error(plot_ord(res))
})

test_that("plot_ord works with unweighted data", {
  t <- c(0, 4, 8, 12, 16, 20)
  set.seed(1)
  conc <- (0.5 * (0.05 * t + (10^(-0.5)) / 0.5))^-2 + rnorm(length(t), 0, 0.05)
  dframe <- data.frame(t, conc)
  res <- det_order(dframe)
  expect_no_error(plot_ord(res))
})

test_that("plot_ord does not give error with no reaction order determined", {
  t <- c(0, 4, 8, 12, 16, 20)
  conc <- c(1, 0.51, 0.51, 0.7, 0.12, 0.07)
  err <- c(0.02, 0.05, 0.04, 0.04, 0.03, 0.02)
  dframe <- data.frame(t, conc, err)
  res <- det_order(dframe)
  expect_equal(res[[6]], NaN, ignore_attr = TRUE)
  expect_no_error(plot_ord(res))
})

test_that("plot_ord works when 0-order model is determined", {
  t <- c(0, 4, 8, 12, 16, 20)
  conc <- c(1, 0.91, 0.80, 0.69, 0.59, 0.51)
  err <- c(0.02, 0.05, 0.04, 0.04, 0.03, 0.02)
  dframe <- data.frame(t, conc, err)
  res <- det_order(dframe)
  expect_no_error(plot_ord(res))
})

test_that("plot_ord works only with ord_res objects", {
  t <- c(0, 4, 8, 12, 16, 20)
  conc <- c(1, 0.51, 0.51, 0.7, 0.12, 0.07)
  err <- c(0.02, 0.05, 0.04, 0.04, 0.03, 0.02)
  pippo <- lm(conc ~ t)
  pippo2 <- nls(conc ~ exp(-k * t), start = list(k = 0.5), weights = 1 / err^2)
  dframe <- data.frame(t, conc, err)
  expect_error(plot_ord(t))
  expect_error(plot_ord(dframe))
  expect_error(plot_ord(pippo))
  expect_error(plot_ord(pippo2))
})

test_that("If estimated model parameters are not significant a particular
          message is displayed - case order==0", {
  t <- c(0, 1, 2, 3, 4, 5)

  set.seed(1)
  conc <- 12 * exp(-0.01 * t) + rnorm(length(t), 0, 1)
  pippo <- data.frame(t, conc, err = abs(rnorm(length(t), 0, 1)))
  expect_no_error(res <- det_order(pippo))
  expect_no_error(plot_ord(res))
  expect_message(
    plot_ord(res),
    "NB: parameters of the regression are not significant!"
  )
})


R/FOMT.R
#' First-Order Multi-Target model regression
#'
#' The function performs a non-linear regression using the first-order
#' multi-target model. The model equation is:
#' \deqn{\frac{S}{S_0}=1-(1-e^{-k\,t})^m}
#' where \eqn{S/S_0} is the fraction of surviving molecules, \eqn{k} is the
#' average number of hits per time unit, \eqn{m} is the number of hits required
#'  to degrade the molecule, and  \eqn{t} is time.
#'
#' @param dtframe A data-frame containing 2 or 3 columns: time, normalized
#' concentration and error (optional), respectively
#'
#' @return Returns the results of the regression as a
#' [nls][stats::nls()] object.
#'
#' @details
#' The FOMT model has been proposed as an alternative to the Weibull equation
#' that is commonly used when the time-dependent behavior of the data
#' significantly deviates from that predicted by standard chemical models.
#'
#' @seealso [FOMTm()][FOMTm()], [par_est_FOMT()]
#' @export
#' @importFrom stats nls
#'
#' @examples
#' t <- c(0, 4, 8, 12, 16, 20)
#' conc <- c(1, 0.98, 0.99, 0.67, 0.12, 0.03)
#' err <- c(0.02, 0.05, 0.04, 0.04, 0.03, 0.02)
#' dframe <- data.frame(t, conc, err)
#' FOMT <- FOMT(dframe)
#' plot(dframe[[1]], dframe[[2]])
#' arrows(dframe[[1]], dframe[[2]] + dframe[[3]],
#'   dframe[[1]], dframe[[2]] - dframe[[3]],
#'   length = 0
#' )
#' newt <- seq(0, 21, by = 0.1)
#' lines(newt, predict(FOMT, newdata = list(t = newt)))
#'
#' dframe1 <- data.frame(t, conc)
#' FOMT1 <- FOMT(dframe1)
#' plot(dframe1[[1]], dframe1[[2]])
#' lines(newt, predict(FOMT1, newdata = list(t = newt)))
#' summary(FOMT)
#' summary(FOMT1)
FOMT <- function(dtframe) {
  t <- dtframe[[1]]
  y <- dtframe[[2]]

  if (length(dtframe[1, ]) == 2) {
    FOMT <- stats::nls(y ~ 1 - (1 - exp(-k * t))^n,
      data = list(y = y, t = t),
      start = par_est_FOMT(t, y)
    )
  } else {
    if (length(dtframe[1, ]) == 3) {
      err <- dtframe[[3]]
      FOMT <- stats::nls(y ~ 1 - (1 - exp(-k * t))^n,
        data = list(y = y, t = t),
        start = par_est_FOMT(t, y),
        weights = 1 / err^2
      )
    } else {
      stop("Unexpected number of columns")
    }
  }
  return(FOMT)
}


#' First-Order Multi-Target model
#'
#' Call the function to return the formula of the Single-Hit Multi-Target
#' model (FOMT):
#' \deqn{1-(1-e^{-k\,t})^n}
#'
#' @param k average number of hits per time unit
#' @param t time
#' @param m minimum number of hits required to degrade the
#' molecule
#'
#' @return Returns calculated values using the formula of FOMT model
#'
#' It can be used inside the function [nls][stats::nls()] as the RHS of the
#' formula.
#' @export
#'
#' @seealso [FOMT()], [par_est_FOMT()], [stats::nls()]
#'
#' @examples
#' t <- seq(0, 100, by = 1)
#' k <- 0.1
#' n <- 200
#' y <- FOMTm(t, k, n)
#' plot(t, y, type = "l")
FOMTm <- function(t, k, m) {
  1 - (1 - exp(-k * t))^m
}

#' First-Order Multi-Target parameter starting values
#'
#' `par_est_FOMT` estimates the starting values of the parameters of the
#' first-order multi-target model from a data-set.
#'
#' @param x A data-frame with time and concentration in the first and second
#' columns,respectively. Alternatively, it could be an array of time and y an
#' array of concentrations.
#' @param y Optional, an array of concentrations. To be inserted only if x is an
#' array.
#'
#' @return The function returns an array with the suggested initial values of
#' parameters.
#' @export
#' @importFrom stats lm coef
#' @seealso [FOMT()], [FOMTm()]
#' @examples
#' t <- seq(0, 30, by = 6)
#' k <- 0.3
#' n <- 40
#' set.seed(100)
#' y <- FOMTm(t, k, n) * (1 + rnorm(length(t), 0, 0.05))
#'
#' nlsFOMT <- nls(y ~ FOMTm(t, k, n),
#'   data = list(y = y, t = t),
#'   start = par_est_FOMT(t, y)
#' )
#' summary(nlsFOMT)
par_est_FOMT <- function(x, y = NULL) {
  if (is.data.frame(x)) {
    t <- x[[1]]
    y <- x[[2]]
  } else {
    t <- x
  }
  t <- t[y < 0.2]
  y <- y[y < 0.2]

  coefs <- stats::coef(
    stats::lm(I(log(y)) ~ t)
  )
  coefs[1] <- exp(coefs[1])
  coefs[2] <- -coefs[2]
  attributes(coefs) <- NULL

  return(c(k = coefs[2], n = coefs[1]))
}


R/data.R
#' First order kinetic data
#'
#' Synthetic data from a first-order kinetic model with k=0.7
#' @usage ord1
#' @format
#' A data frame with 6 rows and 3 columns:
#' \describe{
#'   \item{t}{time}
#'   \item{concentration}{simulated concentration data at each time point}
#'   \item{std.error}{simulated experimental error}
#' }
"ord1"

#' Total CQA with ascorbic acid (FOMT model)
#'
#' Degradation data of 1.2 mM 5-caffeoylquinic acid (5-CQA) in the presence of
#' 1.2 mM of ascorbic acid at 37°C. The data refer to total CQA concentration.
#'
#' @usage fomtdata
#' @format
#' A data frame with 8 rows and 2 columns:
#' \describe{
#'   \item{time_h}{Time in hours}
#'   \item{tCQA_AA}{Normalized concentration of total CQA measured at each
#'   time point}
#' }
#' @source Yusaku N. and Kuniyo I. (2013)
#' *Degradation Kinetics of Chlorogenic
#' Acid at Various pH Values and Effects of Ascorbic Acid and Epigallocatechin
#' Gallate on Its Stability under Alkaline Conditions*,
#' Journal of Agricultural and Food Chemistry, \doi{10.1021/jf304105w},
#'  fig 2D, solid diamonds
"fomtdata"

#' Urfa pepper ascorbic acid degradation (2-nd order)
#'
#' Data describing the degradation kinetics of ascorbic acid during dehydration
#' of Urfa peppers. The peppers were treated with hot air at 55, 65 and 75 °C.
#'
#' @usage urfa
#' @format
#' A data frame with 8 rows and 4 columns:
#' \describe{
#'   \item{time_min}{time in minutes}
#'   \item{AA_55}{normalized concentration of ascorbic acid of Urfa peppers
#'   dehydrated at 55°C}
#'   \item{AA_65}{normalized concentration of ascorbic acid of Urfa peppers
#'   dehydrated at 65°C}
#'   \item{AA_75}{normalized concentration of ascorbic acid of Urfa peppers
#'   dehydrated at 75°C}}
#' @source Ş. Dağhan, A. Yildirim, F. Mehmet Yilmaz, H. Vardin and
#' M. Karaaslan (2018)
#' *The effect of temperature and method of drying on isot (Urfa pepper)
#' and its Vitamin C degradation kinetics*,
#' Italian Journal of Food Science,
#' \doi{10.14674/IJFS-1070},
#' fig 5, hot-air
"urfa"


R/plot.R
#' Plot of phase space and kinetic curve
#'
#' The function plots the results obtained from a [det_order()] function.
#' Two plots are shown: one representing the transformed data in the phase
#' space and the other the kinetic data in the conventional space along with
#' their regression curves.
#'
#' @param ord_res an 'ord_res' object
#'
#' @return Two plots. The first representing the transformed data in the phase
#' space and the other the kinetic data in the conventional space along with
#' their regression curves.
#' Black line represent the best regression curve, whereas green lines show the
#' fits with the reaction order chosen.
#' @export
#'
#' @importFrom graphics par arrows plot abline lines
#' @importFrom stats predict
#' @importFrom methods is
#'
#' @examples
#' t <- c(0, 4, 8, 12, 16, 20)
#' conc <- c(1, 0.51, 0.24, 0.12, 0.07, 0.02)
#' dframe <- data.frame(t, conc)
#' res <- det_order(dframe)
#'
#' plot_ord(res)
plot_ord <- function(ord_res) {
  oldpar<-par(no.readonly = TRUE)
  on.exit(par(oldpar))
  if (methods::is(ord_res, "ord_res")) {
    dflplot <- as.data.frame(ord_res[[1]])

    dfline <- data.frame(
      c(ord_res[[2]]$coefficients[1]),
      c(ord_res[[2]]$coefficients[2])
    )
    names(dfline) <- c("k", "n")

    if (is.nan(ord_res[[6]])) {
      graphics::plot(dflplot$log_x, dflplot$log_dx_dt,
        main = "Phase space",
        xlab = "log(x)", ylab = "log(|dx/dy|)"
      )
      graphics::abline(a = dfline$n, b = dfline$k)
      message <- "The process is not well described by a n-order kinetic model"
      return(list(message))
    }
    dfplot <- as.data.frame(ord_res[[5]])
    if (length(ord_res[[1]][1, ]) == 2) {
      names(dfplot) <- c("x", "y")
    } else {
      names(dfplot) <- c("x", "y", "err")
    }
    xn <- seq(0, max(dfplot$x) * 1.1, by = max(dfplot$x) / 100)
    pn <- stats::predict(ord_res[[4]], newdata = list(t = xn, y0 = dfplot$y[1]))
    dfnls <- data.frame(xn, pn)

    if (ord_res[[6]] == 0) {
      dfnlord <- data.frame(
        c(ord_res[[6]]),
        c(summary(ord_res[[4]])$coefficients[2])
      )
    } else {
      dfnlord <- data.frame(
        c(ord_res[[6]]),
        c(summary(ord_res[[4]])$parameters[1])
      )
    }
    names(dfnlord) <- c("n", "k")
    mm <- min(dflplot$log_x)
    MM <- min(dflplot$log_dx_dt)

    mm1 <- max(dfplot$x)
    MM1 <- max(dfplot$y)

    graphics::par(pty = "s", ask = TRUE)
    graphics::plot(dflplot$log_x, dflplot$log_dx_dt,
      xlab = "log(x)", ylab = "log(|dx/dy|)", main = "Phase space",
      ylim = c(min(dflplot$log_dx_dt), 0), xlim = c(min(dflplot$log_x), 0)
    )
    graphics::abline(a = dfline$k, b = dfline$n)
    graphics::abline(a = log(abs(dfnlord$k)), b = dfnlord$n, col = "red")

    if (length(ord_res[[1]][1, ]) == 4) {
      graphics::arrows(dflplot$log_x, dflplot$log_dx_dt + dflplot$err_log,
        dflplot$log_x, dflplot$log_dx_dt - dflplot$err_log,
        length = 0
      )
      graphics::arrows(dflplot$log_x + dflplot$err_x, dflplot$log_dx_dt,
        dflplot$log_x - dflplot$err_x, dflplot$log_dx_dt,
        length = 0
      )
    }

    graphics::plot(dfplot$x, dfplot$y,
      xlab = "time", ylab = "Concentration", main = "Kinetic degradation"
    )
    graphics::lines(dfnls$xn, dfnls$pn, col = "red")

    if (length(ord_res[[1]][1, ]) == 4) {
      graphics::arrows(dfplot$x, dfplot$y + dfplot$err,
        dfplot$x, dfplot$y - dfplot$err,
        length = 0
      )
    }

    if (!ord_res[[3]]) {
      message("NB: parameters of the regression are not significant!")
    }
  } else {
    stop("Error: the argument is not an object from det_order function")
  }
}


R/ord_res_fun.R
#' Phase space, linearized model
#'
#' Given an `ord_res` object, this function returns the linearized model
#' that best fits the data in the phase space.
#' `ord_res` object can be obtained using the function [det_order()].
#'
#' @param x an `ord_res` object
#'
#' @return Returns a `lm` class object.
#' @seealso [det_order()], [kin_regr()], [results()], [stats::lm()]
#' @export
#' @examples
#' t <- c(0, 4, 8, 12, 16, 20)
#' conc <- c(1, 0.51, 0.24, 0.12, 0.07, 0.02)
#' dframe <- data.frame(t, conc)
#' res <- det_order(dframe)
#'
#' phase_space(res)
phase_space <- function(x) {
  stopifnot(methods::is(x,"ord_res"))
  return(x[[2]])
}

#' Degradation kinetics
#'
#' Returns from an `ord_res` object either the linear or the non-linear
#' regression of the degradation kinetics data.
#'
#' After the
#' analysis in the phase space for the determination of the reaction order,
#' [det_order()] performs either a linear or a non-linear
#' regression of the kinetic data, depending on whether the reaction order is
#' n=0 or n>0,
#' respectively. To access the regression object call `kin_degr`.
#'
#' @param x an `ord_res` object
#'
#' @return Returns either an `nls` or `lm` object based on the regression
#' performed by the function [det_order()].
#' @export
#' @seealso [det_order()], [phase_space()], [results()], [stats::lm()]
#' @examples
#' t <- c(0, 4, 8, 12, 16, 20)
#' conc <- c(1, 0.51, 0.24, 0.12, 0.07, 0.02)
#' dframe <- data.frame(t, conc)
#' res <- det_order(dframe)
#'
#' kin_regr(res)
kin_regr <- function(x) {
  stopifnot(methods::is(x,"ord_res"))
  return(x[[4]])
}

#' Summary of 'ord_res' object
#'
#' Returns the results of the analyses performed by [det_order()] function.
#'
#' The function prints:
#' \enumerate{
#'    \item{the linear regression performed in the phase space, together with
#'    the estimated *n* value and its 95% confidence interval}
#'    \item{a brief conclusion on the results obtained in the phase space
#'    stating which reaction order should be preferred}
#'    \item{the (non-)linear regression performed with parameters associated
#'    statistics. If a non-linear regression has been performed, the most
#'    common goodness-of-fit measures calculated with [goodness_of_fit()]
#'    are printed}
#'    }
#'
#' @param object an 'ord_res' object
#'
#' @return It prints a summary of the analysis in the phase space, the reaction
#' order, and the regression results.
#' @export
#' @importFrom stats coefficients confint
#' @import MASS
#' @seealso [det_order()], [kin_regr()], [phase_space()]
#' @examples
#' t <- c(0, 4, 8, 12, 16, 20)
#' conc <- c(1, 0.51, 0.24, 0.12, 0.07, 0.02)
#' err <- c(0.02, 0.05, 0.04, 0.04, 0.03, 0.02)
#' dframe <- data.frame(t, conc, err)
#' res <- det_order(dframe)
#'
#' results(res)
results <- function(object) {
  cat("\nLinear regression in the phase space:", "\n")
  cat(
    "log(dx/dt)=", round(object[[2]]$coefficients[[2]], 2), "log(x) + (",
    round(object[[2]]$coefficients[[1]], 2), ")\n\n"
  )
  cat("Estimate of n:\n\n")

  ic <- summary(object[[2]])$coefficients[2, ]

  print(ic,
    digits = getOption("digits"), quote = FALSE,
    na.print = "", zero.print = "0",
    right = is.numeric(ic) || is.complex(ic),
    justify = "none"
  )

  cat("\nConfidence interval of n:", "\n")

  ic <- stats::confint(object[[2]])[2, ]

  print(ic,
    digits = getOption("digits"), quote = FALSE,
    na.print = "", zero.print = "0",
    right = is.numeric(ic) || is.complex(ic),
    justify = "none"
  )

  cat("\n")
  if (is.nan(object[[6]])) {
    cat(
      "The process is not appropriately described by a standard n-order",
      "kinetic model. Alternative models should be evaluated.\n"
    )
    proceed <- FALSE
  } else {
    if (round(object[[6]]) == object[[6]] && object[[6]] != 0) {
      cat("Statistical analysis indicates that an order ", object[[6]],
        " degradation kineitc model is likely to describe the data.\n",
        "The null hypothesis H0:\n",
        "\"The process is described by an order ", object[[6]],
        "kinetic model\"\n cannot be rejected.",
        sep = ""
      )
      proceed <- TRUE

      if (!object[[3]]) {
        cat(
          "\n Estimates of model parameters are not significant.",
          "Alternative models should be evaluated"
        )
        return()
      }
    } else {
      if (object[[6]] == 0) {
        proceed <- FALSE
        cat("Estimate of the intercept is significant; but the estimate of",
          " slope is not.\nThe data are likely to be described by an 0-order",
          " kinetic model.",
          sep = ""
        )

        if (!object[[3]]) {
          cat(
            "\n Estimates of model parameters are not significant.",
            "Alternative models should be evaluated."
          )
          return()
        }

        cat(
          "\n\nLinear  regression was performed (0-order kinetics).\n ",
          "\nEstimated k value: \n"
        )
        print(summary(object[[4]]))
        cat("\nConfidence interval of k: \n")
        ic <- -stats::confint(object[[4]])[2, ]
        print(ic,
          digits = getOption("digits"), quote = FALSE,
          na.print = "", zero.print = "0",
          right = is.numeric(ic) || is.complex(ic),
          justify = "none"
        )
      } else {
        cat(
          "The confidence interval of n excludes integer-valued orders.\n",
          "The best estimate n=", object[[6]], "was used to perform the",
          "regression.\n \n"
        )
        proceed <- TRUE
      }
    }
  }

  if (proceed) {
    cat(
      "\n\nNon-linear least squares regression was performed with an order ",
      object[[6]], " kinetic model:\n \n Estimate of k: \n"
    )
    ic <- summary(object[[4]])$parameters
    print(ic,
      digits = getOption("digits"), quote = FALSE,
      na.print = "", zero.print = "0",
      right = is.numeric(ic) || is.complex(ic),
      justify = "none"
    )

    ci2 <- stats::confint(object[[4]])
    cat("Confidence interval of k: \n")
    print(ci2,
      digits = getOption("digits"), quote = FALSE,
      na.print = "", zero.print = "0",
      right = is.numeric(ci2) || is.complex(ci2),
      justify = "none"
    )


    cat("\nGoodness-of-fit:\n")
    tb <- goodness_of_fit(object[[4]])
    print(tb)
    if (is.na(tb[5, 1])) {
      cat(
        "NB: Reduced Chi-squared is not calculated with",
        "unweighted data\n"
      )
    }
  }

  cat("-----------------------------------------------------\n\n")
}


k_value <- function(ordres) {
  if (is(ordres[[4]], "nls")) {
    return(stats::coefficients(ordres[[4]]))
  } else {
    if (is(ordres[[4]], "lm")) {
      return(abs(stats::coefficients(ordres[[4]])))
    } else {
      message(
        "Kinetic model was not found in object. \nProbably the phase",
        "space regression was not significant."
      )
    }
  }
}


n_value <- function(ordres) {
  stopifnot(methods::is(ordres,"ord_res"))
  return(ordres[[6]])
}
